pkg/queue/workers/retention.go

```diff
@@ -0,0 +1,204 @@
+	// These are internal queue internal tasks
+	MaintenanceTaskQueue string = "queue-maintenance"
+
+	// RetentionTask is finished task cleanup type
+	RetentionTask queue.TaskType = "retention"
+)
+
+var (
+	errSerializingHearbeat = errors.New("failed to serialize progress payload while sending heartbeat")
+)
+
+type retentionProgress struct {
+	// Duration of the HTTP request in milliseconds
+	Duration *int64 `json:"duration,omitempty"`
+	// RowsAffected
+	RowsAffected *int64 `json:"rowsAffected,omitempty"`
+	// ErrorMessage contains an error message string if it occurs during the update process
+	ErrorMessage *string `json:"errorMessage,omitempty"`
+}
+
+type retentionSpec struct {
+	// QueueName determine which queue the retention policy applies to
+	QueueName string `json:"queueName"`
+	// TaskType determines which task type the retention policy applies to
+	TaskType queue.TaskType `json:"taskType"`
+	// Status is the required status the task must have to be deleted
+	Status queue.TaskStatus `json:"status"`
+	// Age is the time since finish when the task will be considered for deletion
+	Age time.Duration `json:"age"`
+	// SQL is the actual sql that will be run
+	SQL string `json:"sql"`
+}
+
+// NewRetentionHandler creates a task handler that will clean up old finished tasks
+func NewRetentionHandler(db *sql.DB) TaskHandler {
+	return &retentionHandler{
+		Tracer: tracing.NewTracer("handlers", "RetentionHandler"),
+		db:     db,
+	}
+}
+
+type retentionHandler struct {
+	tracing.Tracer
+	db *sql.DB
+}
+
+func (h *retentionHandler) Process(ctx context.Context, task queue.Task, heartbeats chan<- queue.Progress) (err error) {
+	span, ctx := h.StartSpan(ctx, "Process")
+	defer func() {
+		h.FinishSpan(span, err)
+	}()
+	span.SetTag("task.id", task.ID)
+	span.SetTag("task.queue", task.Queue)
+	span.SetTag("task.type", task.Type)
+	span.SetTag("task.spec", string(task.Spec))
+
+	log := logrus.WithContext(ctx).WithField("type", task.Type)
+	log.Debug("starting retention task")
+
+	var progress retentionProgress
+	defer func() {
+		// we check for errSerializingHearbeat so we don't cause
+		// a recursion call
+		if err == nil || err == errSerializingHearbeat {
+			return
+		}
+		message := err.Error()
+		progress.ErrorMessage = &message
+		_ = sendRetentionProgress(progress, heartbeats)
+	}()
+
+	spec := retentionSpec{}
+	err = json.Unmarshal(task.Spec, &spec)
+	if err != nil {
+		return fmt.Errorf("can not parse the retention task spec %w", err)
+	}
+
+	span.SetTag("spec.queue", spec.QueueName)
+	span.SetTag("spec.taskType", spec.TaskType)
+	span.SetTag("spec.status", spec.Status)
+	span.SetTag("spec.age", spec.Age.String())
+
+	// initial heartbeat
+	err = sendRetentionProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	result, err := h.db.ExecContext(ctx, spec.SQL)
+	duration := time.Since(now).Milliseconds()
+	progress.Duration = &duration
```

Comment:
I thought we would want to have this just in case we see weird behaviors, we can then verify how slow the retention task is

Reply:
Although we should be able to get some of this from the existing task duration metric 🤔